mipengine/controller/node_landscape_aggregator.py

```diff
@@ -0,0 +1,269 @@
+        async_result = await sync_to_async(task.apply_async)(
+            args=args, kwargs=kwargs, connection=connection
+        )
+        while not async_result.ready():
+            total_delay += delay
+            if total_delay > CELERY_TASKS_TIMEOUT:
+                raise TimeoutError(
+                    f"Celery task: {task} didn't respond in {CELERY_TASKS_TIMEOUT}s."
+                )
+            await asyncio.sleep(delay)
+            delay = min(delay * 1.5, 2)  # exponential backoff, max 2 seconds
+        return async_result.get(timeout=CELERY_TASKS_TIMEOUT - total_delay)
+
+    return wrapper
+
+
+def _get_node_socket_addr(node_info: NodeInfo):
+    return f"{node_info.ip}:{node_info.port}"
+
+
+class NodeLandscapeAggregator:
+    def __init__(self):
+        self.keep_updating = True
+
+    async def update(self):
+        """
+        Node Landscape Aggregator(NLA) is a module that handles the aggregation of necessary information,
+        to keep up-to-date and in sync the Node Registry and the Data Model Registry.
+        The Node Registry contains information about the node such as id, ip, port etc.
+        The Data Model Registry contains two types of information, data_models and datasets_location.
+        data_models contains information about the data models and their corresponding cdes.
+        datasets_location contains information about datasets and their locations(nodes).
+        NLA periodically will send requests (get_node_info, get_node_datasets_per_data_model, get_data_model_cdes),
+        to the nodes to retrieve the current information that they contain.
+        Once all information about data models and cdes is aggregated,
+        any data model that is incompatible across nodes will be removed.
+        A data model is incompatible when the cdes across nodes are not identical, except one edge case.
+        The edge case is that the cdes can only contain a difference in the field of 'enumerations' in
+        the cde with code 'dataset' and still be considered compatible.
+        For each data model the 'enumerations' field in the cde with code 'dataset' is updated with all datasets across nodes.
+        Once all the information is aggregated and validated the NLA will provide the information to the Node Registry and to the Data Model Registry.
+        """
+        while self.keep_updating:
+            nodes_addresses = _get_nodes_addresses()
+            nodes_info = await _get_nodes_info(nodes_addresses)
+            local_nodes = [
+                node for node in nodes_info if node.role == NodeRole.LOCALNODE
+            ]
+            datasets_locations = await _get_datasets_locations(local_nodes)
+            datasets_labels = await _get_datasets_labels(local_nodes)
+            node_cdes = await _get_cdes_across_nodes(local_nodes)
+            compatible_data_models = _get_data_models(node_cdes)
+            data_models = _get_data_models_with_valid_datasets(
```

Comment:
What do you mean `get_data_models_with_valid_datasets` ?

Maybe `_get_data_models_with_aggregated_datasets` or something that shows that the dataset information is added. It's not a matter of validity.